MaxText/input_pipeline/_hf_data_processing.py

```diff
@@ -0,0 +1,126 @@
+
+ Licensed under the Apache License, Version 2.0 (the "License");
+ you may not use this file except in compliance with the License.
+ You may obtain a copy of the License at
+
+      https://www.apache.org/licenses/LICENSE-2.0
+
+ Unless required by applicable law or agreed to in writing, software
+ distributed under the License is distributed on an "AS IS" BASIS,
+ WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ See the License for the specific language governing permissions and
+ limitations under the License.
+ """
+
+"""Input pipeline using Huggingface datasets."""
+
+from typing import Optional, Union
+
+import ml_collections
+import jax
+
+from datasets import load_dataset
+from datasets.distributed import split_dataset_by_node
+from transformers import AutoTokenizer
+from torchdata.datapipes.iter import IterableWrapper
+
+from input_pipeline import _hf_operations
+import multihost_dataloading
+
+def get_datasets(
+  config: ml_collections.ConfigDict
+):
+  """Load huggingface dataset"""
+  train_ds = load_dataset(config.dataset_name,
+                            data_dir=config.dataset_dir_hf,
+                            data_files=config.dataset_path,
+                            split="train",
+                            streaming=True,
+                            token=config.hf_access_token)
+  return train_ds, None
+
+def preprocess_dataset(config: ml_collections.ConfigDict,
+                        global_mesh,
+                        train_ds,
+                        add_bos = True,
+                        add_eos = True,
+                        ):
+  """preprocess dataset"""
+  # Set global batch size.
+  global_batch_size_to_load = config.global_batch_size_to_load
+
+  train_iter = preprocessing_pipeline(
+      dataset=train_ds,
+      tokenizer_path=config.tokenizer_path,
+      add_bos=add_bos,
+      add_eos=add_eos,
+      batch_size=global_batch_size_to_load,
+      global_mesh=global_mesh,
+      shuffle=config.enable_data_shuffling,
+      num_epochs=1,
+      pack_examples=True,
+      max_length=config.max_target_length,
+      data_shuffle_seed=config.data_shuffle_seed,
+      access_token=config.hf_access_token,)
+
+  return train_iter, None, None
+
+def preprocessing_pipeline(
+  dataset,
+  tokenizer_path,
+  add_bos: bool,
+  add_eos: bool,
+  batch_size: int,
+  global_mesh,
+  shuffle: bool,
+  num_epochs: Optional[int] = 1,  # only support num_epoch=1 for now
+  pack_examples: bool = True,
+  max_length: int = 512,
+  shift: bool = True,
+  drop_remainder: bool = True,  # does not support drop_remainder
+  data_shuffle_seed = 0,
+  access_token: Union[str | None] = None,
+  prefetch_buffer_size: int = 100,
+):
+  """pipeline for preprocessing"""
+  assert (
+        batch_size % global_mesh.size == 0
+  ), 'Batch size should be divisible number of global devices.'
+
+  dataset = split_dataset_by_node(dataset, world_size=jax.process_count(), rank=jax.process_index())
```

Comment:
@khatwanimohit make sure to sync with Mohit here, this is no longer right.

Reply:
This implemention does not support "expansion_factor_real_data" yet, as I noted in the doc input_pipeline.md. I will add support for expansion_factor_real_data in later PR. (Currently "expansion_factor_real_data" is ignored when dataset_type=hf). This should work for now, @khatwanimohit please also take a look.

Reply:
I don't see a reason why it can't be done in this PR. Its a small change. You'll just have to pass in 
```
dataloading_host_index = process_indices.index(jax.process_index()),
dataloading_host_count = len(process_indices)
```
to `_hf_data_processing.preprocess_dataset`